### PR TITLE
Fixes 422 error due to bad sql building

### LINKED ISSUE #12656: Filtering product by supplier property breaks when enterprise has custom sorting by category set
## Description
When looking at a shop, customer have the ability to filter products by product category, supplier property or product property using the the "tag" under "filter by" on the left hand side of the page.

In the specific scenario where a supplier as set a custom sorting by category, filtering by supplier property doesn't work. When clicking on a supplier property, the "tag" will be selected, but the filtering won't happen and the loading products spinner never disappear.
The bug can be reproduced here : https://staging.openfoodnetwork.org.au/noname-hub/shop when filtering by "Free Range"

This was introduced by the Product refactor

## Expected Behavior
Only product inheriting "Free Range" supplier property should be displayed

filter_bug_expectedl

## Actual Behaviour
All product are shown, no filtering is happening and Loading products spinner doesn't disappear. There is a request returning a 422 error (see bugsnag below)

filter_bug_actual

## Steps to Reproduce

As a enterprise manager 
1. Go to the shop setting, and select "shop preferences" on the left hand side menu
2. Set "Display ordering in shopfront"  to  "By category" and select one or more categories and save change
3. Now select "Properites" on the left hand side menu
4. Add a property
5. Go to the "Products" page and choose a product
6. Add a property to the product and tick "Inherit properties from <shop name>", and update 

As a customer
1. Navigate to the shop page
2.Try to filter by the supplier property added above -> filtering doesn't work

## Error in Coopcircuits

**ActiveRecord::StatementInvalid** in **api/v0/order_cycles#products**
PG::UndefinedColumn: ERROR:  column first_variant.supplier_id does not exist
LINE 4: ...t.product_id JOIN enterprises ON enterprises.id = first_vari...
                                                             ^

[View on Bugsnag](https://app.bugsnag.com/open-food-france/coopcircuits/errors/668d105cf7d2310008d6e4d2?event_id=668dcf8700efc2f7d13c0000&i=gh&m=ci)

Another one on AU prod: 
https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/668cc120963f0b000858abbf?event_id=668dd06e00efe4c55a620000&i=sk&m=nw

## Stacktrace

    app/services/products_renderer.rb:91 - filter
    app/services/products_renderer.rb:41 - products
    app/services/products_renderer.rb:20 - products_json
    app/controllers/api/v0/order_cycles_controller.rb:26 - products

[View full stacktrace](https://app.bugsnag.com/open-food-france/coopcircuits/errors/668d105cf7d2310008d6e4d2?event_id=668dcf8700efc2f7d13c0000&i=gh&m=ci)

*Created by Coopcircuits Admin via Bugsnag*

## Workaround
Can still browse product it just less convenient to not be able to filter.

## Severity
bug-s3: a feature is broken but there is a workaround

### PR DESCRIPTION
**Pull request by cyrillefr**
#### What? Why?

- Closes #12656

When some properties are defined at the enterprise level, there is a 422 return code as well as a "Products loading ..." displayed when customer filters products.

Reason is a bug in SQL Query building.
First part of query use `supplier_properties` parameter, but not second part, that can leads to mismatch between the 2 parts:
https://github.com/openfoodfoundation/openfoodnetwork/blob/dfea0cd805c555b84dce163f9ed83842a5b5b992/app/services/order_cycles/distributed_products_service.rb#L37-L40

The 2nd bit:
https://github.com/openfoodfoundation/openfoodnetwork/blob/dfea0cd805c555b84dce163f9ed83842a5b5b992/app/services/order_cycles/distributed_products_service.rb#L81-L83

#### What should we test?
See § "Steps to Reproduce" from #12656

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [X] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by cyrillefr** [comment]
Hello @rioug ,

I uploaded a bug here in the `order` method (the `||` is not good).

I have working on the `order` method locally, it looks a bit like that

```ruby
elsif sorting_with_supplier_properties?(supplier_properties)
  order_by_property = distributor.producer_properties 
    .map { |pp| "producer_properties.property_id=#{pp.property_id} DESC" } 
    .join(", ")
     "#{order_by_property}, spree_products.name ASC, spree_products.id ASC"
```
```ruby
def sorting_with_supplier_properties?(supplier_properties)
  supplier_properties && distributor.producer_properties.present?
end
```

I am not sure whether my use of `distributor.producer_properties` is good( I have seen it used in specs).
But I realise now that may be I should make another `if` in the `relation_by_sorting` method, to use
`producer_properties.property_id` in the `select` and in the `group by`.

What do you think ?

**Comment by rioug** [comment]
If I am following correctly what you are saying, I think you are trying to something we don't need. There is no need to order by `producer_properties`.

I think what we need to do here, it's to get rid of the `supplier_properties` flag (and the `||` ) in `relation_by_sorting` and update the the query when we are sorting "by_category", we need to include the `first_variant.supplier_id` in the select. Doing that allows us to use `supplier_property_join` when sorting "by_category" without blowing up. Does that make sense ? I have a feeling that might not work but it's a good place to start.

Good job on cleaning up  `order !`

**Comment by rioug** [comment]
As a first step you should try to add the missing spec in here : 
https://github.com/openfoodfoundation/openfoodnetwork/blob/dfea0cd805c555b84dce163f9ed83842a5b5b992/spec/services/products_renderer_spec.rb#L139
We need the scenario where `preferred_shopfront_taxon_order` is set on the distributor and we are filtering by producer properties.

**Comment by cyrillefr** [comment]
Hello @rioug ,

Thanks for the tip.
It seems I handled this issue from the wrong end with this supplier_properties parameter ....
So, I removed it from the `relation_by_sorting`.
I added `supplier_id` in the 2nd part of the method, and it works.
I also added one spec.

Checks have failed, but I cannot reproduce one and the 2nd fail is a redis issue ...
Could you please rerun the checks ? It should be ok

**Review by rioug — APPROVED** [review]
Awesome, thanks for fixing this @cyrillefr :pray: ! I had a feeling it wasn't going to be this simple but I am glad to be proven wrong :smile:

File: spec/services/products_renderer_spec.rb [hunk 1]
```diff
@@ -155,6 +155,19 @@
...
+        it "filters products with producer properties when sorting is enabled" do
+          allow(distributor).to receive(:preferred_shopfront_taxon_order) {
+            "#{fruits.id},#{cakes.id}"
+          }
+          fruits_supplier.producer_properties.create!({ property_id: property_organic.id,
+                                                        value: '1', position: 1 })
+          search_param = { q: { "with_variants_supplier_properties" => [property_organic.id] } }
+          products_renderer = ProductsRenderer.new(distributor, order_cycle, customer, search_param)
+
+          products = products_renderer.send(:products)
```
Comment (dacook) [inline spec/services/products_renderer_spec.rb:168]: According to our rubocop rules, this needs to use the `__send__` method.
Rubocop was skipping this file, but we recently enabled it (after this branch was made).
I will go ahead and correct it.
```suggestion
          products = products_renderer.__send__(:products)
```

Comment (dacook) [reply spec/services/products_renderer_spec.rb:168]: It's worth noting that usage of `send`/`__send__` is generally not good practice, and we should be testing the result of the public methods. So next time, we should try to do that. This is fine for now though.

**Review by dacook — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Thanks so much @cyrillefr ,

I could verify the bug is fixed indeed in staging. Before this PR, the bug is reproducible:

image


After this PR the bug is gone, and the correct results are returned:
image

Merging!
Thank you.

**Pull request merged**